File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-4.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if len(s) < 1:
            return len(s)
        l, r = 0, 1

        maxLen = 1
        hashSet = set()
        hashSet.add(s[l])
        while r < len(s):
            while s[r] in hashSet:
                hashSet.remove(s[l])
                l += 1
            hashSet.add(s[r])
            
            maxLen = max(maxLen, len(hashSet))
            r += 1
           
                
        return maxLen

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-4.cpp
class Solution {
public:
    int characterReplacement(string s, int k) {
        unordered_map<char, int> m;

        int left = 0; 

        int maxCharacter = 0;
        int maxLen = 0;

        for (int right = 0; right < s.size(); right++) {
            m[s[right]]++;
            maxCharacter = max(maxCharacter, m[s[right]]);

            while ((right - left + 1) - maxCharacter > k) {
                m[s[left]]--;
                left++;
            }

            maxLen = max(maxLen, right - left + 1);
        }

         return maxLen;
    }
};
